src/abstractdataframe/show.jl

```diff
@@ -68,23 +68,46 @@ if VERSION < v"1.5.0-DEV.261" || VERSION < v"1.5.0-DEV.266"
     end
 end
 
-"""Return compact string representation of type T"""
-function compacttype(T::Type, maxwidth::Int=8, initial::Bool=true)
-    maxwidth = max(8, maxwidth)
+# memoize calls to compacttype
+const TYPE2STRING_DICT = Dict{Any, String}()
+
+# use this lock on compacttype caller side to ensure thread safety
+const TYPE2STRING_LOCK = Threads.ReentrantLock()
+
+"""
+    compacttype(T::Type, maxwidth::Int=8, initial::Bool=true)
 
+Return compact string representation of type T.
+This function is not thread safe. Use TYPE2STRING_LOCK to ensure thread safety.
+"""
+function compacttype(T::Type, maxwidth::Int=8, initial::Bool=true)
     T === Any && return "Any"
     T === Missing && return "Missing"
 
+    if haskey(TYPE2STRING_DICT, (T, maxwidth, initial))
+        return TYPE2STRING_DICT[(T, maxwidth, initial)]
+    end
```

Comment:
This avoids doing the lookup twice:
```suggestion
    match = get(TYPE2STRING_DICT, (T, maxwidth, initial), nothing)
    match !== nothing && return match
```